### PR TITLE
Add Plugin for UTXO tracking by account.

### PR DESCRIPTION
**Pull request by jsolman**
This plugin facilitates sending NEO or GAS for arbitrary addresses by providing an RPC method `getunspents` for retrieving the unspent UTXOs. The unspent UTXOs are stored indexed by account script hash in an additional LevelDB database whose path is configurable in the plugin settings. This plugin was migrated from my initial PR here: neo-project/neo/pull/569

The format of the response data is compatible with the format returned from the neoscan v1 API get_balance` call.

### CODE REVIEW

**Comment by jsolman** [comment]
I'll be testing this by re-syncing MainNet from the latest chain file and will report here with the results from my testing.

**Comment by vncoelho** [comment]
Building and testing it now.

**Comment by vncoelho** [comment]
I got this error but need double check, maybe it is my configuration.

```cs
[ERROR][03/11/2019 20:24:58][Thread 0004][akka://NeoSystem/user/$a] Error while creating actor instance of type Neo.Ledger.Blockchain with 2 args: (Neo.NeoSystem,Neo.Persistence.LevelDB.LevelDBStore)
Cause: [akka://NeoSystem/user/$a#360886005]: Akka.Actor.ActorInitializationException: Exception during creation ---> System.TypeLoadException: Error while creating actor instance of type Neo.Ledger.Blockchain with 2 args: (Neo.NeoSystem,Neo.Persistence.LevelDB.LevelDBStore) ---> System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.AggregateException: One or more errors occurred. (Object reference not set to an instance of an object.) ---> System.NullReferenceException: Object reference not set to an instance of an object.
   at Neo.Plugins.RpcSystemAssetPlugin.OnCommit(Snapshot snapshot)
   at Neo.Ledger.Blockchain.Persist(Block block)
   --- End of inner exception stack trace ---
   at Neo.Ledger.Blockchain.Persist(Block block)
   at Neo.Ledger.Blockchain..ctor(NeoSystem system, Store store)
   --- End of inner exception stack trace ---
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at System.Reflection.RuntimeConstructorInfo.Invoke(BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.RuntimeType.CreateInstanceImpl(BindingFlags bindingAttr, Binder binder, Object[] args, CultureInfo culture, Object[] activationAttributes)
   at Akka.Actor.Props.ActivatorProducer.Produce()
   at Akka.Actor.Props.NewActor()
   --- End of inner exception stack trace ---
   at Akka.Actor.Props.NewActor()
   at Akka.Actor.ActorCell.CreateNewActorInstance()
   at Akka.Actor.ActorCell.<>c__DisplayClass109_0.<NewActor>b__0()
   at Akka.Actor.ActorCell.UseThreadContext(Action action)
   at Akka.Actor.ActorCell.NewActor()
   at Akka.Actor.ActorCell.Create(Exception failure)
   --- End of inner exception stack trace ---
   at Akka.Actor.ActorCell.Create(Exception failure)
   at Akka.Actor.ActorCell.SysMsgInvokeAll(EarliestFirstSystemMessageList messages, Int32 currentState)
```

**Comment by jsolman** [comment]
@vncoelho Maybe it is trying to persist the genesis block before the plugin is loaded. I'll check into it.

**Comment by vncoelho** [comment]
Maybe the folder can be just  `RpcSystemAsset`.

**Comment by vncoelho** [comment]
Sorry for not checking it in detail, @jsolman. I will also try to check it more tomorrow.

**Comment by jsolman** [comment]
> Maybe the folder can be just RpcSystemAsset.

Maybe `RpcSystemAssetTracker`

**Comment by vncoelho** [comment]
@jsolman, it works for scripthash, right? Not base58?

**Comment by vncoelho** [comment]
```cs
[ERROR][03/11/2019 21:16:49][Thread 0003][akka://NeoSystem/user/$a] Error while creating actor instance of type Neo.Ledger.Blockchain with 2 args: (Neo.NeoSystem,Neo.Persistence.LevelDB.LevelDBStore)
Cause: [akka://NeoSystem/user/$a#1123270816]: Akka.Actor.ActorInitializationException: Exception during creation ---> System.TypeLoadException: Error while creating actor instance of type Neo.Ledger.Blockchain with 2 args: (Neo.NeoSystem,Neo.Persistence.LevelDB.LevelDBStore) ---> System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.NullReferenceException: Object reference not set to an instance of an object.
   at Neo.Plugins.RpcSystemAssetTrackerPlugin.ResetBatch()
   at Neo.Plugins.RpcSystemAssetTrackerPlugin.OnPersist(Snapshot snapshot, IReadOnlyList`1 applicationExecutedList)
   at Neo.Ledger.Blockchain.Persist(Block block)
   at Neo.Ledger.Blockchain..ctor(NeoSystem system, Store store)
   --- End of inner exception stack trace ---
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at System.Reflection.RuntimeConstructorInfo.Invoke(BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.RuntimeType.CreateInstanceImpl(BindingFlags bindingAttr, Binder binder, Object[] args, CultureInfo culture, Object[] activationAttributes)
   at Akka.Actor.Props.ActivatorProducer.Produce()
   at Akka.Actor.Props.NewActor()
   --- End of inner exception stack trace ---
   at Akka.Actor.Props.NewActor()
   at Akka.Actor.ActorCell.CreateNewActorInstance()
   at Akka.Actor.ActorCell.<>c__DisplayClass109_0.<NewActor>b__0()
   at Akka.Actor.ActorCell.UseThreadContext(Action action)
   at Akka.Actor.ActorCell.NewActor()
   at Akka.Actor.ActorCell.Create(Exception failure)
   --- End of inner exception stack trace ---
   at Akka.Actor.ActorCell.Create(Exception failure)
   at Akka.Actor.ActorCell.SysMsgInvokeAll(EarliestFirstSystemMessageList messages, Int32 currentState)
```
This log is more precise now, @jsolman

File: RpcSystemAssetTracker/UserUnspentCoinOutputsKey.cs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        {
+            if (ReferenceEquals(null, obj)) return false;
+            if (ReferenceEquals(this, obj)) return true;
+            if (obj.GetType() != this.GetType()) return false;
+            return Equals((UserUnspentCoinOutputsKey) obj);
+        }
+
+        public int CompareTo(UserUnspentCoinOutputsKey other)
+        {
+            if (ReferenceEquals(this, other)) return 0;
```
Comment (vncoelho) [inline RpcSystemAssetTracker/UserUnspentCoinOutputsKey.cs]: Are these checks here in CompareTo correct?
In the other cases of equal it returns false when compared to null.

In this sense, compare other to null and turn 1 means that a not null object is better, right?

File: RpcSystemAssetTracker/RpcSystemAssetTrackerPlugin.cs [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+
+        public void PreProcess(HttpContext context, string method, JArray _params)
+        {
+        }
+
+        public JObject OnProcess(HttpContext context, string method, JArray _params)
+        {
+            if (method != "getunspents") return null;
+
+            UInt160 scriptHash = UInt160.Parse(_params[0].AsString());
```
Comment (vncoelho) [inline RpcSystemAssetTracker/RpcSystemAssetTrackerPlugin.cs]: Can we add support for address here, such as https://github.com/neo-project/neo-plugins/pull/43/commits/782d77726ee5b60bd894937e18b399dc0c8a157f?

**Comment by jsolman** [comment]
> @jsolman, it works for scripthash, right? Not base58?

I'll make it work for both.

File: RpcSystemAssetTracker/RpcSystemAssetTrackerPlugin.cs (hunk 2 above)
Comment (jsolman) [reply RpcSystemAssetTracker/RpcSystemAssetTrackerPlugin.cs]: done.

File: RpcSystemAssetTracker/UserUnspentCoinOutputsKey.cs (hunk 1 above)
Comment (jsolman) [reply RpcSystemAssetTracker/UserUnspentCoinOutputsKey.cs]: Returning 1 from compareTo means the objects are not equal, and that a non-null object is greater. I believe this implementation is correct; however, it should not be called with a null other object in this code anyway. It is defensive.

**Comment by jsolman** [comment]
I need to change the plugins so they can override a method that will be called when all plugins have finished loading. The `ImportBlocks` needs to start its work after all plugins are loaded for example. Also, the genesis block needs to not be persisted until plugins are loaded.

**Comment by jsolman** [comment]
Blocks are importing now and I'll report back on how testing goes.
This change is required: neo-project/neo/pull/633

**Comment by jsolman** [comment]
Testing after syncing with MainNet looks good. Keep in mind that to use this you need:
1. To be using Neo 2.10.1 (currently master at or above commit https://github.com/neo-project/neo/commit/432a3f6b5ca4389f7a2221fc845edf770e32b0f7)
2. The updated ImportBlocks plugin: https://github.com/neo-project/neo-plugins/pull/53

Also, to be safe currently, do not stop the import blocks process when importing from chain file by using the `exit` command or `control+c`. I am going to improve shutdown to capture control+c and shutdown cleanly, but currently exit will hang while importing blocks, and interrupting could possibly stop after a block had been committed to the main LevelDB database and before the UTXOs are committed to this LevelDB database, leaving them in an inconsistent state.

As an improvement, I could likely have it persist the block number in LevelDB when this plugin commits so that it can check if it didn't persist the latest block on start-up and handle it that way.

**Comment by jsolman** [comment]
44d358a seems to have hurt performance significantly more than expected. Looking into why.

File: RpcSystemAssetTracker/RpcSystemAssetTrackerPlugin.cs [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+            var unspentsCache = new DbCache<UserSystemAssetCoinOutputsKey, UserSystemAssetCoinOutputs>(
+                _db, null, null, SystemAssetUnspentCoinsPrefix);
+
+            return GenerateUtxoResponse(scriptHash, startingToken, maxIterations, unspentsCache);
+        }
+
+        public JObject OnProcess(HttpContext context, string method, JArray _params)
+        {
+            if (_shouldTrackUnclaimed && method == "getunclaimedspents")
+                return ProcessGetUnclaimedSpents(GetScriptHashFromParam(_params[0].AsString()));
```
Comment (jsolman) [inline RpcSystemAssetTracker/RpcSystemAssetTrackerPlugin.cs]: I expect to adjust this API to get available claim transactions. I just did this quickly intermediately.

Comment (vncoelho) [reply RpcSystemAssetTracker/RpcSystemAssetTrackerPlugin.cs]: This PR is of great impact for the ecosystem and applications in the upper layer.
Congrulations.

Comment (jsolman) [reply RpcSystemAssetTracker/RpcSystemAssetTrackerPlugin.cs]: Output should match neoscan get_claimable api results.

Comment (jsolman) [reply RpcSystemAssetTracker/RpcSystemAssetTrackerPlugin.cs]: I implemented the getclaimable rpc call as mentioned.

**Comment by jsolman** [comment]
I have not done any testing yet after adding the last commit that adds the RPC method to get the claimable spent transactions. I will test it tomorrow and also likely add a method for getting the total unclaimed gas available for an address to fill the last gap.

**Comment by jsolman** [comment]
The API for getting all the claimable transactions is working as expected from my testing with MainNet. Now all that remains is implementing an operation to get total claimable amount, and it should be done.

**Comment by jsolman** [comment]
Ok, support for all operations needed for claiming GAS now are in and seem to be working as expected.

**Comment by jsolman** [comment]
Maybe we can create a 2.10.1 branch as you suggested elsewhere; I will not merge this as it references a CI NuGet build.

**Pull request merged**